### ISSUE TITLE
[INTEGRATION][SPARK] `InMemoryRelation` not supported as in/out dataset

### ISSUE THREAD

**Issue by wslulciuc** [OpenLineage/OpenLineage#610]
Let's add support for [`InMemoryRelation`](https://spark.apache.org/docs/1.3.1/api/java/org/apache/spark/sql/columnar/InMemoryRelation.html) (see also InMemoryRelation [internals](https://jaceklaskowski.gitbooks.io/mastering-spark-sql/content/spark-sql-LogicalPlan-InMemoryRelation.html) internals) as an in/out dataset for a job run when traversing the spark logic plan. For more details, see `spark_unknown` facet below.

```
{
  "spark_unknown": {
    "inputs": [
      {
        "description": {
          "id": 1,
          "@class": "org.apache.spark.sql.execution.columnar.InMemoryRelation",
          "origin": {
            "id": 20
          },
          "output": [...],
          "schema": {
            "type": "struct",
            "fields": [...]
          },
          "resolved": true,
          "streaming": false,
          "traceEnabled": false,
          "allAttributes": {
            "id": 21,
            "attrs": []
          },
          "outputOrdering": [],
          "validConstraints": [],
          "canonicalizedPlan": false
        },
        "inputAttributes": [],
        "outputAttributes": [...],
    "_producer": "https://github.com/OpenLineage/OpenLineage/tree/0.6.1/integration/spark",
    "_schemaURL": "https://openlineage.io/spec/1-0-2/OpenLineage.json#/$defs/RunFacet"
  },
  "spark_version": {
    "_producer": "https://github.com/OpenLineage/OpenLineage/tree/0.6.1/integration/spark",
    "_schemaURL": "https://openlineage.io/spec/1-0-2/OpenLineage.json#/$defs/RunFacet",
    "spark-version": "3.0.1.17-***",
    "openlineage-spark-version": "0.6.1"
  },
  "spark.logicalPlan": {
    "plan": [
      {
        "class": "org.apache.spark.sql.catalyst.plans.logical.OverwritePartitionsDynamic",
        "query": 0,
        "table": [
          {
            "class": "org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation",
            "output": [ ... ],
    "_producer": "https://github.com/OpenLineage/OpenLineage/tree/0.6.1/integration/spark",
    "_schemaURL": "https://openlineage.io/spec/1-0-2/OpenLineage.json#/$defs/RunFacet"
  }
}
```

**Comment by collado-mike** [comment]
I don't think I'd call this a bug - I think there's some decision making around how in-memory datasets should be reported. 

It's tricky to consider, as there's no proper naming convention for the dataset. What's the scheme? `memory://`? and the dataset name? If we name all of the in-memory datasets the same, every Spark job that loads data from memory will look like it reads from the same dataset as every other job, which is clearly not the case.

We can do something like `mem://<fully_qualified_name_of_job>`. Even then, though, there's significant risk of showing lineage where there shouldn't be. You can imagine two dataframes in a spark application that read different arrays and are joined together, but the lineage graph might show a single dataset being read and joined with itself...

**Comment by collado-mike** [comment]
So after discussing today, I realized I mistook `InMemoryRelation` for a `LocalRelation` - the former is a cached view of a `SparkPlan`, while the latter is a dataset that originates in memory (e.g., `SparkContext.parallelize`).

I was able to write a unit test to exercise this case (and validate that we don't correctly handle it) - 
```java
  @Test
  public void testCacheReadFromFileWriteToJdbc(@TempDir Path writeDir, SparkSession spark)
      throws InterruptedException, TimeoutException, IOException {
    Path testFile = writeTestDataToFile(writeDir);
    Path tableOneDir = writeDir.resolve("table1");

    spark.read().json("file://" + testFile.toAbsolutePath().toString())
        .createOrReplaceTempView("raw_json");
    spark.sql("CACHE TABLE cached_json AS SELECT * FROM raw_json WHERE age > 100");
    Dataset<Row> df = spark.sql("SELECT * FROM cached_json");
    df.collect();

    Path sqliteFile = writeDir.resolve("sqlite/database");
    sqliteFile.getParent().toFile().mkdir();
    df.write().parquet(tableOneDir.toAbsolutePath().toString());

    // wait for event processing to complete
    StaticExecutionContextFactory.waitForExecutionEnd();

    ArgumentCaptor<OpenLineage.RunEvent> lineageEvent =
        ArgumentCaptor.forClass(OpenLineage.RunEvent.class);

    Mockito.verify(SparkAgentTestExtension.OPEN_LINEAGE_SPARK_CONTEXT, times(18))
        .emit(lineageEvent.capture());
    List<OpenLineage.RunEvent> events = lineageEvent.getAllValues();
    ObjectAssert<RunEvent> completionEvent =
        assertThat(events)
            .filteredOn(e -> e.getEventType().equals(RunEvent.EventType.COMPLETE))
            .isNotEmpty()
            .filteredOn(e -> !e.getInputs().isEmpty())
            .isNotEmpty()
            .filteredOn(e -> !e.getOutputs().isEmpty())
            .isNotEmpty()
            .filteredOn(e -> e.getOutputs().stream().anyMatch(o -> o.getOutputFacets() != null))
            .isNotEmpty()
            .first();
    completionEvent
        .extracting(RunEvent::getInputs, InstanceOfAssertFactories.list(InputDataset.class))
        .hasSize(1)
        .first()
        .hasFieldOrPropertyWithValue("namespace", "file")
        .hasFieldOrPropertyWithValue("name", testFile.toAbsolutePath().getParent().toString());

    completionEvent
        .extracting(RunEvent::getOutputs, InstanceOfAssertFactories.list(OutputDataset.class))
        .hasSize(1)
        .first()
        .hasFieldOrPropertyWithValue("namespace", "file")
        .hasFieldOrPropertyWithValue("name", tableOneDir.toAbsolutePath().toString())
        .satisfies(
            d -> {
              // Spark rowCount metrics currently only working in Spark 3.x
              if (spark.version().startsWith("3")) {
                assertThat(d.getOutputFacets().getOutputStatistics())
                    .isNotNull()
                    .hasFieldOrPropertyWithValue("rowCount", 2L);
              }
            });
  }
```

The output of this is a series of SQL execution jobs
```
CreateViewCommand `raw_json`, false, true, LocalTempView
   +- Relation[age#7L,name#8] json

CacheTableCommand [cached_json], SELECT * FROM raw_json WHERE age > 100, false
   +- 'Project [*]
      +- 'Filter ('age > 100)
         +- 'UnresolvedRelation [raw_json], [], false

Aggregate [count(1) AS count#28L]
+- Project
   +- InMemoryRelation [age#11L, name#12], StorageLevel(disk, memory, deserialized, 1 replicas)
         +- *(1) Filter (isnotnull(age#7L) AND (age#7L > 100))
            +- FileScan json [age#7L,name#8] Batched: false, DataFilters: [isnotnull(age#7L), (age#7L > 100)], Format: JSON, Location: InMemoryFileIndex[file:/var/folders/x6/s53ym0hn1sv9sjjyxjdld7t40000gn/T/junit7230691869285414069/..., PartitionFilters: [], PushedFilters: [IsNotNull(age), GreaterThan(age,100)], ReadSchema: struct<age:bigint,name:string>

InsertIntoHadoopFsRelationCommand file:/var/folders/x6/s53ym0hn1sv9sjjyxjdld7t40000gn/T/junit7230691869285414069/table1, false, Parquet, [path=/var/folders/x6/s53ym0hn1sv9sjjyxjdld7t40000gn/T/junit7230691869285414069/table1], ErrorIfExists, [age, name]
+- InMemoryRelation [age#74L, name#75], StorageLevel(disk, memory, deserialized, 1 replicas)
      +- *(1) Filter (isnotnull(age#7L) AND (age#7L > 100))
         +- FileScan json [age#7L,name#8] Batched: false, DataFilters: [isnotnull(age#7L), (age#7L > 100)], Format: JSON, Location: InMemoryFileIndex[file:/var/folders/x6/s53ym0hn1sv9sjjyxjdld7t40000gn/T/junit7230691869285414069/..., PartitionFilters: [], PushedFilters: [IsNotNull(age), GreaterThan(age,100)], ReadSchema: struct<age:bigint,name:string>
```

The latter plan is the one we really want to evaluate, as it tells us that we are reading from the JSON file and writing out to the parquet file. Unfortunately, the `InMemoryRelation` only holds a reference to the cached `SparkPlan`, not the original `QueryExecution` or its `LogicalPlan`. The framework for visiting plans only works for `LogicalPlan` nodes, not `SparkPlan` nodes.

One option we could try would be to output events for the `CreateViewCommand` and/or the `CacheTableCommand`. However, that only works for datasets cached using SQL, as in this test case. If I change the Spark code to the following, the logical plan events are totally different:
```java
    Dataset<Row> cachedDf = spark.read().json("file://" + testFile.toAbsolutePath().toString())
        .filter("age > 100")
        .persist(StorageLevel.MEMORY_ONLY());
    cachedDf.collect();

    Path sqliteFile = writeDir.resolve("sqlite/database");
    sqliteFile.getParent().toFile().mkdir();
    cachedDf.write().parquet(tableOneDir.toAbsolutePath().toString());
```

```
InMemoryRelation [age#7L, name#8], StorageLevel(memory, deserialized, 1 replicas)
   +- *(1) Filter (isnotnull(age#7L) AND (age#7L > 100))
      +- FileScan json [age#7L,name#8] Batched: false, DataFilters: [isnotnull(age#7L), (age#7L > 100)], Format: JSON, Location: InMemoryFileIndex[file:/var/folders/x6/s53ym0hn1sv9sjjyxjdld7t40000gn/T/junit3622079451878339523/..., PartitionFilters: [], PushedFilters: [IsNotNull(age), GreaterThan(age,100)], ReadSchema: struct<age:bigint,name:string>

InsertIntoHadoopFsRelationCommand file:/var/folders/x6/s53ym0hn1sv9sjjyxjdld7t40000gn/T/junit3622079451878339523/table1, false, Parquet, [path=/var/folders/x6/s53ym0hn1sv9sjjyxjdld7t40000gn/T/junit3622079451878339523/table1], ErrorIfExists, [age, name]
+- InMemoryRelation [age#7L, name#8], StorageLevel(memory, deserialized, 1 replicas)
      +- *(1) Filter (isnotnull(age#7L) AND (age#7L > 100))
         +- FileScan json [age#7L,name#8] Batched: false, DataFilters: [isnotnull(age#7L), (age#7L > 100)], Format: JSON, Location: InMemoryFileIndex[file:/var/folders/x6/s53ym0hn1sv9sjjyxjdld7t40000gn/T/junit3622079451878339523/..., PartitionFilters: [], PushedFilters: [IsNotNull(age), GreaterThan(age,100)], ReadSchema: struct<age:bigint,name:string>
```
Here, we never get an event that tells us what the `InMemoryRelation` contains except in the underlying `SparkPlan`.

**Comment by pawel-big-lebowski** [comment]
Caching datasets in Spark is an extremely popular mechanism for storing in memory intermediate datasets, dictionaries that reused within a single Spark run, etc. Unfortunately, current OpenLineage integration does not support it. It is failing on the as simple job as the one below:

```
spark.createDataFrame([
    {'a': 1, 'b': 2},
    {'a': 3, 'b': 4}
]).write.saveAsTable('temp')

cached = spark.read.table("temp").cache()       # create cached dataset
cached.count()                                  # call an action on cached dataset
cached.select('a').write.saveAsTable('target')  # target table lineage should be aware of temp table
```
An OL event generated, when saving `target` table, does not contain information about data origin which is `temp` table. This is because `InMemoryRelation` which is contained within a logical plan and contains `SparkPlan` (a physical plan used to create cached dataset) instead of a logical plan. I also tried to make use of `SparkPlan.logicalLink` method which returns `the logical plan this plan is linked to`, but it returns the an empty optional. 

What probably happens here, is:
 * When evaluating count on `cached` dataset a logical plan is rewritten into a physical plan of a dataset which is then attached to a dataset. 
 * An action triggered to write `target` tables is based on a copy version of `SparkPlan` object and a copy does not contain `LogicalPlan` used to create it. 

The solution to that problem requires providing the link between the run that creates in-memory dataset (which knows how dataset was created) and the run that makes use of it. 

What if we introduced a concept of in-memory input/output datasets like 
`spark-memory://{spark-session-id}/{spark-plan-id}`?

A dataset unique name needs to contain spark session identifier as dataset resides only in memory of each spark session. In that case we could a create an OL event with output in-memory dataset and then providing its identifier when being used. This would require joining information from two events at the backend side.

I did not have a chance to verify it but this looks likes a further direction to follow.
